### PR TITLE
Remove folding

### PR DESCRIPTION
**Pull request by Maijin**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the documentation and the [rizin book](https://github.com/rizinorg/rizinbook) with the relevant information (if needed)

**Detailed description**

Removing folding feature that's adding complexity in core .c files (disasm.c, visual.c)

### CODE REVIEW

**Comment by XVilka** [comment]
I am not sure about this one - folding is useful in the visual interface.

**Comment by Maijin** [comment]
imho It's adding complexity, it's not working, tested properly and almost no users knows about it. I think may be good to add it back later once disasm.c/visual.c is decluttered.

**Comment by thestr4ng3r** [comment]
Personally, I am ok with removing since I never used this. Don't know about others though.

**Comment by XVilka** [comment]
I just checked with:
```
rizin /bin/ls
[0x00006b60]> aaa
[0x00006b60]> afF
[0x00006b60]> Vp
```
And it works:

image

**Comment by Maijin** [comment]
Yeah but in visual as soon as there is more complexity everytime there is a new addition, folding needs to be taken in the code/display. It impacts scrolling/mouse wheels gets sluggish, visual display etc. You can see the number of files impacted, where it should barely touch a couple, imho not good enough to be part of a cleaned-up disasm.c / visual.c.

**Comment by kazarmy** [comment]
Perhaps only basic blocks should be folded, and a folded function is a special case of a folded superblock? Anyway, it would be great if we could see broken folding examples.

**Comment by thestr4ng3r** [comment]
I think instead of folding individual blocks and functions, it might make more sense to be able to make groups of blocks and then fold/unfold such groups. So you can get a more high-level graph with maybe some complex parts being collapsed into a single folded block. That would indeed require rewriting this from scratch.

**Comment by Maijin** [comment]
Just an example - https://asciinema.org/a/J4gA4dvXbwDWfp8GoLZXOrwYT folding a function with z, then it's showing weird .int32 and scroll seems broken afterwards :/

**Comment by wargio** [comment]
tbh, looks really weird.

**Comment by kazarmy** [comment]
Yeah, the code probably needs an answer to the question "What happens if you seek into the middle of a folded block/function?" among others.

**Comment by thestr4ng3r** [comment]
Considering all of this, I think it's better to remove.

**Comment by ret2libc** [comment]
I agree with removing this. As @XVilka said, it works, yes, but as soon as you do any other thing it's kinda unusable imo. So +1 from me.

**Review by XVilka — APPROVED** [review]
LGTM to remove then.  I created [an issue](https://github.com/rizinorg/rizin/issues/203) for implementing this feature from scratch. Let's merge this PR.

**Comment by XVilka** [comment]
Just please fix this warning/error:
```
../librz/core/visual.c: In function ‘rz_core_visual_disasm_down’:
../librz/core/visual.c:4056:22: error: variable ‘f’ set but not used [-Werror=unused-but-set-variable]
  RzAnalysisFunction *f = NULL;
                      ^
```

File: librz/core/visual.c [hunk 1]
```diff
@@ -4112,26 +4056,22 @@ RZ_API void rz_core_visual_disasm_down(RzCore *core, RzAsmOp *op, int *cols) {
 	RzAnalysisFunction *f = NULL;
 	f = rz_analysis_get_fcn_in (core->analysis, core->offset, 0);
```
Comment (wargio) [inline librz/core/visual.c]: i think you need to remove this and the previous line.

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/analysis/serialize_analysis.c [hunk 2]
```diff
@@ -13,13 +13,13 @@
  *
  * /
  *   /blocks
- *     0x<addr>={size:<ut64>, jump?:<ut64>, fail?:<ut64>, traced?:true, folded?:true, colorize?:<ut32>,
```
Comment (ret2libc) [inline librz/analysis/serialize_analysis.c:16]: @thestr4ng3r I think we can consider the project format changeable until 1.0.0, right?

File: librz/core/agraph.c [hunk 3]
```diff
@@ -4362,18 +4347,6 @@ RZ_API int rz_core_visual_graph(RzCore *core, RzAGraph *g, RzAnalysisFunction *_
 				agraph_toggle_callgraph (g);
 			}
 			break;
-		case 'Z':
```
Comment (ret2libc) [inline librz/core/agraph.c:4365]: Actually.. Is there anything wrong with the folding on the ascii-graph? I think it works well enough... or not?

**Review by ret2libc — APPROVED** [review]

File: librz/core/agraph.c (hunk 3 above)
Comment (Maijin) [reply librz/core/agraph.c:4365]: @ret2libc Doesn't work see:

image

Comment (ret2libc) [reply librz/core/agraph.c:4365]: Did you press z while in graph mode or while in visual mode?

Comment (Maijin) [reply librz/core/agraph.c:4365]: while in graph mode

Comment (ret2libc) [reply librz/core/agraph.c:4365]: ahhh sorry, I was doing the `z`, not `Z`. Ok then!

**Pull request merged**